client/core/core.go

```diff
@@ -2618,42 +2523,154 @@ func (c *Core) Register(form *RegisterForm) (*RegisterResult, error) {
+		}
+		// Check the DEX server's signature.
+		// If this was a legacy registration, we'll be setting the pubkey here. For
+		// v2 registrations, we're overwriting the existing key, but whatever. We
+		// already checked that they're identical a few lines up.
+		msg := regRes.Serialize()
+		dc.acct.dexPubKey, err = checkSigS256(msg, regRes.DEXPubKey, regRes.Sig)
+		if err != nil {
+			return nil, false, newError(signatureErr, "%s pubkey error: %v", dc.acct.host, err)
+		}
+		// Fresh account registration success.
+		return
+	}
+
+	// Registration error could be AccountExistsError or AccountSuspendedError.
+	var msgErr *msgjson.Error
+	if errors.As(err, &msgErr) {
+		switch msgErr.Code {
+		case msgjson.AccountSuspendedError:
+			inProcessReg.acctSuspended = true
+			acctType := "HD"
+			if len(encKey) == 0 {
+				acctType = "Legacy"
+			}
+			c.log.Infof("%s account key for %s was reported as suspended.", acctType, dc.acct.host)
+			return inProcessReg, false, nil
+
+		case msgjson.AccountExistsError:
+			// This is now account recovery, which is great news since we don't
+			// have to pay the fee. Server provides fee coin as the error message.
+			c.log.Infof("%s is reporting that this account already exists. Skipping fee payment.", dc.acct.host)
+			dc.acct.feeCoin, err = hex.DecodeString(msgErr.Message)
+			if err != nil {
+				return nil, true, fmt.Errorf("error decoding fee coin ID from message %q", msgErr.Message)
+			}
+
+			dcrID, _ := dex.BipSymbolID(regFeeAssetSymbol)
+			cid, _ := asset.DecodeCoinID(dcrID, dc.acct.feeCoin)
+			c.log.Infof("Recovered paid account for %s. Fee previously paid with %s", dc.acct.host, cid)
+
+			err := c.db.CreateAccount(&db.AccountInfo{
+				Host:         dc.acct.host,
+				Cert:         dc.acct.cert,
+				LegacyEncKey: encKeyLegacy,
+				EncKeyV2:     encKey,
+				DEXPubKey:    dc.acct.dexPubKey,
+				FeeCoin:      dc.acct.feeCoin,
+			})
+			if err != nil {
+				// Shouldn't let the client trade with this server if we can't store
+				// the acct details to DB. Since this is not a fresh registration with
+				// a randomly generated acct key, the client can restore this account
+				// after fixing any db issues before proceeding to trade.
+				return nil, true, fmt.Errorf("Error saving restored account: %w", err)
+			}
+
+			err = c.db.AccountPaid(&db.AccountProof{
+				Host:  dc.acct.host,
+				Stamp: 54321,
+				Sig:   []byte("RECOVERY SIGNATURE"),
+			})
+			if err != nil {
+				return nil, true, fmt.Errorf("Error marking recovered account as paid: %w", err)
+			}
+
+			dc.acct.markFeePaid()
+
+			err = c.authDEX(dc)
+			if err != nil {
+				return nil, true, fmt.Errorf("error authorizing pre-paid account: %v", err)
+			}
+
+			if !dc.broadcastingConnect() { // dc might already be listening for messages if called from Register which creates a non-temporary dc.
+				atomic.StoreUint32(&dc.reportingConnects, 1)
```

Comment:
A compare and swap here would be appropriate instead of using the `broadcastingConnect` helper